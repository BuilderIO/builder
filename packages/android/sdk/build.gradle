buildscript {
    ext {
        compose_version = '1.1.0-beta01'
    }

    ext.kotlin_version = '1.6.10'
    repositories { 
        mavenCentral() 
        google() 
    }

    dependencies {
        classpath "org.jetbrains.kotlin:kotlin-serialization:$kotlin_version"
        classpath 'com.google.gms:google-services:4.3.13'
    }
}// Top-level build file where you can add configuration options common to all sub-projects/modules.
plugins {
    id 'com.android.application' version '7.2.1' apply false
    id 'com.android.library' version '7.2.1' apply false
    id 'org.jetbrains.kotlin.android' version '1.5.31' apply false
    id 'org.jetbrains.kotlin.plugin.serialization' version '1.6.10'
}

task clean(type: Delete) {
    delete rootProject.buildDir
}

allprojects {
  repositories {
    google()
  }
}
