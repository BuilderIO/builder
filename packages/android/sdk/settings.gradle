pluginManagement {
    repositories {
        gradlePluginPortal()
        google()
        mavenCentral()
    }
}
dependencyResolutionManagement {
    repositoriesMode.set(RepositoriesMode.PREFER_SETTINGS)
    repositories {
        google()
        mavenCentral()
    }
}
rootProject.name = "My Application"
include ':app'
