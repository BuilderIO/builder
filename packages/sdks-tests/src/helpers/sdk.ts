import { z } from 'zod';

export type Sdk =
  | 'oldReact'
  | 'qwik'
  | 'react'
  | 'reactNative'
  | 'rsc'
  | 'solid'
  | 'svelte'
  | 'vue'
  | 'angular';

const serverNameEnumValues = [
  'react-native',
  'solid',
  'solid-start',
  'qwik-city',
  'next-pages-dir',
  'react',
  'remix',
  'hydrogen',
  'next-app-dir-client',
  'next-app-dir',
  'vue',
  'nuxt',
  'svelte',
  'sveltekit',
  'angular',
  'angular-ssr',
  'gen1-react',
  'gen1-remix',
  'gen1-next',
] as const;
const ServerNameEnum = z.enum(serverNameEnumValues);
export type ServerName = z.infer<typeof ServerNameEnum>;

const envServerName = !process.env.SERVER_NAME
  ? []
  : process.env.SERVER_NAME === 'all'
    ? serverNameEnumValues
    : process.env.SERVER_NAME.split(',');

export const serverNames = envServerName.map(str => ServerNameEnum.parse(str)) ?? [];

if (serverNames.length === 0) {
  throw new Error(
    'SERVER_NAME is required. Please provide a comma-separated list of server names to run.'
  );
}

/**
 * Map of server names to SDKs.
 */
export const SDK_MAP: Record<ServerName, Sdk> = {
  'react-native': 'reactNative',
  solid: 'solid',
  'solid-start': 'solid',
  'qwik-city': 'qwik',
  'next-pages-dir': 'react',
  react: 'react',
  remix: 'react',
  hydrogen: 'react',
  'next-app-dir-client': 'react',
  vue: 'vue',
  nuxt: 'vue',
  svelte: 'svelte',
  sveltekit: 'svelte',
  'next-app-dir': 'rsc',
  'gen1-react': 'oldReact',
  'gen1-next': 'oldReact',
  'gen1-remix': 'oldReact',
  angular: 'angular',
  'angular-ssr': 'angular',
};
