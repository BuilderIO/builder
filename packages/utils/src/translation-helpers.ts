import { BuilderContent, BuilderElement } from '@builder.io/sdk';
import traverse from 'traverse';
import omit from 'lodash/omit';
const localizedType = '@builder.io/core:LocalizedValue';

export type TranslateableFields = {
  [key: string]: {
    instructions?: string;
    value: string;
  };
};

export function getTranslateableFields(
  content: BuilderContent,
  sourceLocaleId: string,
  defaultInstructions: string
) {
  const results: TranslateableFields = {};

  let { blocks, blocksString, state, ...customFields } = content.data!;

  if (typeof blocks === 'undefined') {
    blocks = JSON.parse(blocksString);
  }

  // metadata [content's localized custom fields]
  traverse(customFields).forEach(function (el) {
    if (this.key && el && el['@type'] === localizedType) {
      results[`metadata.${this.path.join('#')}`] = {
        instructions: el.meta?.instructions || defaultInstructions,
        value: el[sourceLocaleId] || el.Default,
      };
    }
  });

  // blocks
  traverse(blocks).forEach(function (el) {
    // TODO: localized custom components inputs
    if (el && el.id && el.component?.name === 'Text' && !el.meta?.excludeFromTranslation) {
      results[`blocks.${el.id}#text`] = {
        value: el.component.options.text,
        instructions: el.meta?.instructions || defaultInstructions,
      };
    }
  });

  return results;
}

export function appLyTranslation(
  content: BuilderContent,
  translation: TranslateableFields,
  locale: string
) {
  let { blocks, blocksString, state, ...customFields } = content.data!;
  if (typeof blocks === 'undefined') {
    blocks = JSON.parse(blocksString);
  }

  traverse(customFields).forEach(function (el) {
    const path = this.path?.join('#');
    if (translation[`metadata.${path}`]) {
      this.update({
        ...el,
        [locale]: translation[`metadata.${path}`].value,
      });
    }
  });

  traverse(blocks).forEach(function (el) {
    // TODO: localized custom components inputs
    if (
      el &&
      el.id &&
      el.component?.name === 'Text' &&
      !el.meta?.excludeFromTranslation &&
      translation[`blocks.${el.id}#text`]
    ) {
      this.update({
        ...el,
        meta: {
          ...el.meta,
          translated: true,
        },
        bindings: {
          ...el.bindings,
          'component.options.text': `state.translation['blocks.${el.id}#text'][state.locale || 'Default'] || \`${el.component.options.text}\``,
        },
      });
    }
  });

  const translationState = Object.keys(translation).reduce((acc, key) => {
    if (key.startsWith('blocks.')) {
      return {
        ...acc,
        [key]: {
          '@type': localizedType,
          ...content.data!.state?.translation?.[key],
          [locale]: translation[key].value,
        },
      };
    }

    return acc;
  }, {});

  content.data!.state = {
    ...content.data!.state,
    translation: translationState,
  };

  content.data = {
    ...omit(content.data, 'blocksString'),
    blocks,
    ...customFields,
  };

  return content;
}
