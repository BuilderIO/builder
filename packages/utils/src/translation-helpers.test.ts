import { BuilderContent } from '@builder.io/sdk';
import { appLyTranslation, getTranslateableFields } from './translation-helpers';

test('getTranslateableFields from content to match snapshot', async () => {
  const content: BuilderContent = {
    data: {
      title: {
        '@type': '@builder.io/core:LocalizedValue',
        'en-US': 'Hello',
        Default: 'Test',
      },
      blocks: [
        {
          meta: {
            instructions: 'This is a mobile only element',
          },
          id: 'block-id',
          '@type': '@builder.io/sdk:Element',
          component: {
            name: 'Text',
            options: {
              text: 'test',
            },
          },
        },
      ],
    },
  };
  const result = getTranslateableFields(
    content,
    'en-US',
    'Visit https://builder.io/fiddle/... for more details'
  );
  expect(result).toMatchSnapshot();
});

test('applyTranslation from content to match snapshot', async () => {
  const content: BuilderContent = {
    data: {
      title: {
        '@type': '@builder.io/core:LocalizedValue',
        'en-US': 'Hello',
        Default: 'Test',
      },
      blocks: [
        {
          id: 'block-id',
          '@type': '@builder.io/sdk:Element',
          component: {
            name: 'Text',
            options: {
              text: 'test',
            },
          },
        },
      ],
    },
  };

  const translations = {
    'metadata.title': { value: 'salut' },
    'blocks.block-id#text': { value: 'block-id' },
  };

  const result = appLyTranslation(content, translations, 'fr-FR');
  expect(result).toMatchSnapshot();
});
