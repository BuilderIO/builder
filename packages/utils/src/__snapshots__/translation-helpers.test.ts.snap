// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applyTranslation from content to match snapshot 1`] = `
Object {
  "data": Object {
    "blocks": Array [
      Object {
        "@type": "@builder.io/sdk:Element",
        "bindings": Object {
          "component.options.text": "state.translation['blocks.block-id#text'][state.locale || 'Default'] || \`test\`",
        },
        "component": Object {
          "name": "Text",
          "options": Object {
            "text": "test",
          },
        },
        "id": "block-id",
        "meta": Object {
          "translated": true,
        },
      },
    ],
    "state": Object {
      "translation": Object {
        "blocks.block-id#text": Object {
          "@type": "@builder.io/core:LocalizedValue",
          "fr-FR": "block-id",
        },
      },
    },
    "title": Object {
      "@type": "@builder.io/core:LocalizedValue",
      "Default": "Test",
      "en-US": "Hello",
      "fr-FR": "salut",
    },
  },
}
`;

exports[`getTranslateableFields from content to match snapshot 1`] = `
Object {
  "blocks.block-id#text": Object {
    "instructions": "This is a mobile only element",
    "value": "test",
  },
  "metadata.title": Object {
    "instructions": "Visit https://builder.io/fiddle/... for more details",
    "value": "Hello",
  },
}
`;
